File: src/main/Python/boj_step16/boj9012.py
import sys
cin = sys.stdin.readline

def isVPS(s):
    L = []
    for c in s:
        if c == '(': L.append(c)
        elif c == ')':
            if len(L) == 0: return False
            else: L.pop()
    return len(L) == 0

for _ in range(int(cin())):
    print('YES' if isVPS(cin()) else 'NO')

File: src/main/Cpp/boj_step16/boj9012.cpp
#include <iostream>
#include <stack>
#include <functional>
using namespace std;

int main() {
    function isVPS = [&] (stack<char> st, string ps) -> string {
        for (const auto & p : ps) {
            if (p == '(')
                st.push(p);
            else {
                if (st.empty())
                    return "NO";

                st.pop();
            }
        }

        return (st.empty()) ? "YES" : "NO";
    };

    int T;
    cin >> T;

    for (int i = 0; i < T; i++) {
        stack<char> st;
        string ps;
        cin >> ps;
        cout << isVPS(st, ps) << endl;
    }
}